### ISSUE TITLE
Behavior around tracking log files

### ISSUE THREAD

**Issue by Benjamintf1** [rsyslog/rsyslog#2537]
Start: We want to track files in a particular folder(ideally recursively) and as they are created they should be registered and watched.

Step 1: Create /test folder(this exists at start time) and /test/job folder
`mkdir -p /test/job`

Step 2: Configure module 
```
module(load="imfile" mode="inotify")

input(type="imfile"
     Tag="imfile"
     File="/test/*/*"
     Facility="user"
     Severity="info")
```

Step 3: append to new file /test/job/log.log
`
echo $RANDOM >> /test/job/log.log 
#This is to get around log de-duplication
`

Step 4: Expect to see logs in forwarded logs, don't see logs
`There are no logs on the forwarded syslog stream`

Step 5: restart rsyslog
`service rsyslog restart`

Step 6: Echo again to log file, see output(also you should see previous output on restart)
`
echo $RANDOM >> /test/job/log.log 
`
produces log line.

Step 7: Create a folder job2 in test
`mkdir /test/job2`

Step 8: Create log file in job2
`echo $RANDOM >> /test/job2/log.log`

Step 9: See new log file correctly watched/outputs a log line

Step 10: restart rsyslog
`service rsyslog restart`

Step 11: continue logging in logfile in job2 and see the output is still correct
`echo $RANDOM >> /test/job2/log.log`

Step 12: Try and create a new logfile in job2, see it not output
`echo $RANDOM >> /test/job2/log2.log`

Step 13: restart rsyslog
`service rsyslog restart`

Step 14: See all log2 outputs being streamed on restart, and log2 is now watched
`echo $RANDOM >> /test/job2/log2.log`

Step 15: Create a directory in job2, lets try adding a few
`mkdir -p /test/job2/one/two/three/four/five`

Step 16: create log file in leaf and see that it is tracked/output occurs
`echo $RANDOM >> /test/job2/one/two/three/four/five/log.log`

Step 17: restart rsyslog 
`service rsyslog restart`

Step 18: See file is no longer tracked. 
`echo $RANDOM >> /test/job2/one/two/three/four/five/log.log`

Step 19: create log file directly in test, and see it is not tracked
`echo $RANDOM >> /test/log.log`

Now that actual behavior is observed, there comes a few questions:

1. Is * supposed to recurse? If so, why do we see it not work on restart like step 18? If not, why do we see it start tracking files way down the file path in step 16 when creating them? 

1. Why are new files not being tracked? Based on the semantics, I would expect files in the job directory to be tracked, but they are not until a restart occurs, conversely, all files in job2 are tracked until a restart occurs, in which case they stop being tracked. 

What I consider bugs:

1. File tracking behavior should absolutely in no condition be effected negatively by restarts. If a file should be tracked by a file glob it should always and consistently be tracked regardless of when it was created(when the containing folder was created) and when rsyslog was restarted, especially not if the behavior is backwards(functions poorly if folder was created before restart). (This is not even delving into race conditions, these events happen seconds-minutes apart)

2. File tracking should have consistent and reliable documentation on what file patterns mean for the product. (What should /test/*/* track? What should it not?)

Environment Setup:
The current environment setup is a vm created on gcp with the a bosh created stemcell version 3469.1

`https://s3.amazonaws.com/bosh-core-stemcells/google/bosh-stemcell-3469.1-google-kvm-ubuntu-trusty-go_agent.tgz`

This is then configured with forwarding rules and a imfile module. 

Any questions/Thoughts?

**Comment by rgerhards** [comment]
Known issues, see https://github.com/rsyslog/rsyslog/projects/4

If you are interested, I am looking for real-world testers for the refactored code when it is done.

**Comment by Benjamintf1** [comment]
We're definitely able to test the new version to ensure it works when done. Hopefully the specific test cases/uses can help with this issue(I wasn't sure whether the issue was covered, or only partially covered, or not covered by other issues).

**Comment by rgerhards** [comment]
@Benjamintf1 PR https://github.com/rsyslog/rsyslog/pull/2533 is now almost ready and should solve all of your issues. The code is complete, except that I need to re-enable full testbench. If you could give the new code try, that would be fantastic.

**Comment by Benjamintf1** [comment]
We attempted to test this, and ran into difficulties with imfile. As is we are able to see rsyslog handling logs from logger, and we built rsyslog using ./configure --enable-imfile, however now we seem to be getting no logs from the imfile module. We suspect we may have built it wrong in some subtle way, but don't see any errors or other indication of how we might build it right.

We're on Ubuntu Trusty. Is there simple way for us to get to the point of being able to run the new code for testing purposes? Is there some set of flags we're missing? A build artifact somewhere we could use or something?

**Comment by rgerhards** [comment]
There are now daily packages available in the ppa.

**Comment by Benjamintf1** [comment]
Excellent, thank you. We were able to test with this, and it resolves much of our concern; things work as we'd expect.

Some explicit clarification in the documentation about how wildcards work, specifically how they seem to follow the file-globbing pattern and doesn't handle recursive watches, would be nice.

We'll open a separate issue about our recursive-watch/** feature request.
